slither/utils/upgradeability.py

```diff
@@ -0,0 +1,463 @@
+from typing import Optional
+from slither.core.declarations import (
+    Contract,
+    Structure,
+    Enum,
+    SolidityVariableComposed,
+    SolidityVariable,
+    Function,
+)
+from slither.core.solidity_types import (
+    ElementaryType,
+    ArrayType,
+    MappingType,
+    UserDefinedType,
+)
+from slither.core.variables.local_variable import LocalVariable
+from slither.core.variables.local_variable_init_from_tuple import LocalVariableInitFromTuple
+from slither.core.variables.state_variable import StateVariable
+from slither.analyses.data_dependency.data_dependency import get_dependencies
+from slither.core.variables.variable import Variable
+from slither.core.expressions.literal import Literal
+from slither.core.expressions.identifier import Identifier
+from slither.core.expressions.call_expression import CallExpression
+from slither.core.expressions.assignment_operation import AssignmentOperation
+from slither.core.cfg.node import Node, NodeType
+from slither.slithir.operations import (
+    Assignment,
+    Index,
+    Member,
+    Length,
+    Binary,
+    Unary,
+    Condition,
+    NewArray,
+    NewStructure,
+    NewContract,
+    NewElementaryType,
+    SolidityCall,
+    Delete,
+    EventCall,
+    LibraryCall,
+    InternalDynamicCall,
+    HighLevelCall,
+    LowLevelCall,
+    TypeConversion,
+    Return,
+    Transfer,
+    Send,
+    Unpack,
+    InitArray,
+    InternalCall,
+)
+from slither.slithir.variables import (
+    TemporaryVariable,
+    TupleVariable,
+    Constant,
+    ReferenceVariable,
+)
+from slither.tools.read_storage.read_storage import SlotInfo, SlitherReadStorage
+
+
+# pylint: disable=too-many-locals
+def compare(v1: Contract, v2: Contract) -> dict:
+    """
+    Compares two versions of a contract. Most useful for upgradeable (logic) contracts,
+    but does not require that Contract.is_upgradeable returns true for either contract.
+
+    Args:
+        v1: Original version of (upgradeable) contract
+        v2: Updated version of (upgradeable) contract
+
+    Returns: dict {
+        "missing-vars-in-v2": list[Variable],
+        "new-variables": list[Variable],
+        "tainted-variables": list[Variable],
+        "new-functions": list[Function],
+        "modified-functions": list[Function],
+        "tainted-functions": list[Function]
+    }
+    """
+
+    order_vars1 = [v for v in v1.state_variables if not v.is_constant and not v.is_immutable]
```

Comment:
Here we want to use `state_variables_ordered` https://github.com/crytic/slither/blob/master/slither/core/declarations/contract.py#L363

`state_variables` does not maintain the order of declarations, and dont include private variables:
https://github.com/crytic/slither/blob/64b2e0c063cf93fb479481abc432371404629f32/slither/core/declarations/contract.py#L345-L349